File: go-gotest/hello_varar.steps.go
package example

import "github.com/varar-dev/varar/go/varar"

func registerHelloVar(s *varar.Steps[Ctx]) {
	s.Stimulus("I greet {string}", func(ctx Ctx, name string) (Ctx, error) {
		ctx.Greeting = "Hello, " + name + "!"
		return ctx, nil
	})

	s.Sensor("the greeting should be {string}", func(ctx Ctx, expected string) (string, error) {
		return ctx.Greeting, nil
	})

	s.Stimulus("expression `{int}+{int}`", func(ctx Ctx, a, b int) (Ctx, error) {
		ctx.Result = a + b
		return ctx, nil
	})

	s.Sensor("evaluate to `{int}`", func(ctx Ctx, expected int) (int, error) {
		return ctx.Result, nil
	})
}

File: rust-cargotest/src/varar/hello_varar.steps.rs
use super::Ctx;
use ::varar::Steps;

pub fn register(s: &mut Steps<Ctx>) {
    s.stimulus("I greet {string}", |ctx: Ctx, name: String| {
        Ok(Ctx {
            greeting: format!("Hello, {name}!"),
            ..ctx
        })
    });

    s.sensor("the greeting should be {string}", |ctx: Ctx, _expected: String| Ok(ctx.greeting));

    s.stimulus("expression `{int}+{int}`", |ctx: Ctx, a: i64, b: i64| {
        Ok(Ctx {
            result: a + b,
            ..ctx
        })
    });

    s.sensor("evaluate to `{int}`", |ctx: Ctx, _expected: i64| Ok(ctx.result));
}
